prereq_flowchart/classinfo.py

```diff
@@ -0,0 +1,232 @@
+import re
+from enum import IntEnum
+from prereq_flowchart.depts import ALL_DEPTS
+import pickle
+import pprint
+
+CLASSINFO_URL: str = "http://classinfo.umn.edu/?subject={department}&json=1"
+
+
+class Semester(IntEnum):
+    SPRING = 0
+    SUMMER = 1
+    FALL = 2
+
+
+@dataclass(order=True)
+class Term:
+    year: int
+    semester: Semester
+
+    @classmethod
+    def from_str(cls, str_repr: str) -> "Term":
+        term_str, year_str = str_repr.split(" ")
+        year = int(year_str)
+
+        match term_str:
+            case "Spring":
+                term = Semester.SPRING
+            case "Summer":
+                term = Semester.SUMMER
+            case "Fall":
+                term = Semester.FALL
+            case _:
+                raise ValueError(f"Unknown term: {term!r}")
+
+        return cls(year=year, semester=term)
+
+
+@dataclass(frozen=True)
+class CourseNumber:
+    # e.g the CSCI, EE, CEGE part of a course number
+    department: str
+    # the 1001, 1301W, 1402H part of a course number
+    number: str
+
+
+@dataclass
+class Prerequisite:
+    catalog_number: CourseNumber
+    concurrent_allowed: bool = False
+
+
+Credits: TypeAlias = None | float | Tuple[float, float]
+
+
+@dataclass
+class Course:
+    catalog_number: CourseNumber
+    course_title: str
+    # the number of credits that this class is worth
+    credits: Credits
+    # other courses that you have to take before you can take this course
+    prerequisites: list[Prerequisite]
+    is_elective: bool = False
+
+
+def deduplicate_class_listings(
+    class_listings: dict[str, dict[Any, Any]]
+) -> list[dict[Any, Any]]:
+    """Given courses from several terms, keep only the course from the latest term"""
+    course_number_entry_keys: dict[CourseNumber, str] = {}
+    last_seen_course_term: dict[CourseNumber, Term] = {}
+    # key is catalog number
+    for key, course in class_listings.items():
+        course_num = CourseNumber(
+            department=course["Subject"], number=course["Catalog Number"]
+        )
+        course_term = Term.from_str(course["Term"])
+
+        if (
+            course_num not in last_seen_course_term
+            or last_seen_course_term[course_num] < course_term
+        ):
+            last_seen_course_term[course_num] = course_term
+            course_number_entry_keys[course_num] = key
+
+    return [
+        class_listings[entry_key] for entry_key in course_number_entry_keys.values()
+    ]
+
+
+def get_prerequisites_string(class_json: dict[Any, Any]) -> str:
+    """Get a prerequisite string out of a classinfo course desc"""
+
+    def find_rough_string(class_json: dict[Any, Any]) -> str:
+        if "Prerequisites" in class_json:
+            return str(class_json["Prerequisites"])
+        elif "Course Catalog Description" in class_json:
+            desc_str = class_json["Course Catalog Description"]
+            matches = re.match(r"^.+[pP]rereq(?:(?:uisites)|(?:s))?:?(.+)$", desc_str)
```

Comment:
i think we should come up with something that's easier to manipulate if new cases arise (like just a list of strings to be checked against)

Reply:
i think we're gonna completely rewrite the way that prerequisites are extracted from course descriptions in the future (#10), so i am ok with keeping this the way it is for the purposes of this PR